web3/middleware/validation.py

```diff
@@ -27,20 +41,72 @@ def validate_chain_id(web3, chain_id):
         )
 
 
+def check_extradata_length(val):
+    if not isinstance(val, (str, int, bytes)):
+        return val
+    result = HexBytes(val)
+    if len(result) > MAX_EXTRADATA_LENGTH:
+        raise ValidationError(
+            "The field extraData is %d bytes, but should be %d. "
+            "It is quite likely that you are connected to a POA chain. "
+            "Refer "
+            "http://web3py.readthedocs.io/en/latest/middleware.html#geth-style-proof-of-authority "
+            "for more details. The full extraData is: %r" % (
+                len(result), MAX_EXTRADATA_LENGTH, result
+            )
+        )
+    return val
+
+
 def transaction_normalizer(transaction):
     return dissoc(transaction, 'chainId')
 
 
-def validation_middleware(make_request, web3):
-    transaction_validator = apply_formatters_to_dict({
-    })
+def transaction_param_validator(web3):
+    transactions_params_validators = {
+        'chainId': apply_formatter_if(
+            # Bypass `validate_chain_id` if chainId can't be determined
+            lambda _: is_not_null(web3.net.chainId),
+            validate_chain_id(web3)
```

Comment:
Cool, this looks like a good solution that can stay in place for the long term :+1: